### PR TITLE
Allow clusters with multiple node types

### PR DESCRIPTION
**Pull request by adderan**
Not ready for review yet - this is kind of a mess after resolving merge conflicts. Opening PR to see test results.

### CODE REVIEW

**Comment by adderan** [comment]
Looks like the failure is in JobServiceTest, which is also failing non-deterministically on the master branch. So I think this PR is ready for review whenever.

File: src/toil/batchSystems/mesos/batchSystem.py [hunk 1]
```diff
@@ -149,11 +149,16 @@ def __init__(self, config, maxCores, maxMemory, maxDisk):
...
+        self.ignoredNodes = set()
+
         self._startDriver()
 
     def setUserScript(self, userScript):
         self.userScript = userScript
 
+    def ignoreNode(self, nodeAddress):
+        self.ignoredNodes.add(nodeAddress)
+
```
Comment (joelarmstrong) [inline src/toil/batchSystems/mesos/batchSystem.py]: I like this ignored node stuff. But could we add a method to *remove* ignored nodes to AbstractBatchSystem, and call it from clusterScaler at the same time it removes a node from its ignoredNodes set? (clusterScaler.py line 597 as of right now.) It's not out of the question that an instance could be marked to be ignored, terminate, then hours later a new instance might be launched that happens to get the same IP.

File: src/toil/leader.py [hunk 2]
```diff
@@ -572,22 +572,24 @@ def getNumberOfJobsIssued(self, preemptable=None):
...
 
-    def getNumberAndAvgRuntimeOfCurrentlyRunningJobs(self):
+    def getNumberAndAvgRuntimeOfCurrentlyRunningJobs(self, nodeShape=None):
         """
         Returns a tuple (x, y) where x is number of currently running jobs and y
         is the average number of seconds (as a float)
         the jobs have been running for.
         """
         runningJobs = self.batchSystem.getRunningBatchJobIDs()
+        runningJobs = {jobID:runningJobs[jobID] for jobID in runningJobs if getNodeShapeForJob(jobShape=self.jobBatchSystemIDToJobNode[jobID], nodeShapes=self.provisioner.nodeShapes) == nodeShape}
```
Comment (joelarmstrong) [inline src/toil/leader.py]: getNodeShapeForJob isn't defined anywhere.

Comment (joelarmstrong) [inline src/toil/leader.py]: Similarly, the jobBatchSystemIDToJobNode attribute doesn't appear to be set anywhere either. Actually, it looks like this entire function is dead code, which is why these undefined variables don't cause an error. Can we remove this? (and, potentially, the version in clusterScalerTest?)

File: src/toil/provisioners/clusterScaler.py [hunk 3]
```diff
@@ -626,18 +691,14 @@ def _getInfo(allMesosNodes, ip):
...
-            raise RuntimeError('Provisioner could not terminate all autoscaled nodes. There are '
-                               '%s nodes left in the cluster, %s of which were statically provisioned' % (totalNodes, len(self.getStaticNodes(preemptable)))
-                               )
-        elif totalNodes < len(self.scaler.provisioner.getStaticNodes(preemptable)):  # ignore static nodes
-            raise RuntimeError('Provisioner incorrectly terminated statically provisioned nodes.')
-
+        for nodeShape in self.nodeShapes:
+            preemptable = nodeShape.preemptable
+            nodeType = self.nodeShapeToType[nodeShape]
+            totalNodes = self.setNodeCount(nodeType=nodeType, numNodes=0, preemptable=preemptable, force=True)
```
Comment (joelarmstrong) [inline src/toil/provisioners/clusterScaler.py]: pyflakes says `local variable 'totalNodes' is assigned to but never used`. It looks like the call to self.setNodeCount is necessary, but maybe we can just ignore the return value?

File: src/toil/provisioners/clusterScaler.py [hunk 4]
```diff
@@ -468,49 +512,63 @@ def setNodeCount(self, numNodes, preemptable=False, force=False):
...
+                self.ignoredNodes.add(node.privateIP)
+                self.scaler.leader.batchSystem.ignoreNode(node.privateIP)
+
+            if not force:
+                # Filter out nodes with jobs still running. These
+                # will be terminated in _removeIgnoredNodes later on
+                # once all jobs have finished, but they will be ignored by
+                # the batch system and cluster scaler from now on
+                nodesToTerminate = [(node,nodeInfo) for (node,nodeInfo) in nodesToTerminate if nodeInfo is not None and nodeInfo.workers < 1]
+            #nodesToTerminate = nodesToTerminate[:numNodes]
```
Comment (joelarmstrong) [inline src/toil/provisioners/clusterScaler.py]: stray commented out code, looks like this can be removed

**Comment by adderan** [comment]
Jenkins, test this please.

**Comment by adderan** [comment]
Jenkins, test this please.

**Comment by jessebrennan** [comment]
@adderan I've run into this $#!+ before where run_tests.py is out of sync with your actual tests and it's a pain in the @$$. I think this should fix it. 

Also you may want to rebase #1773.

**Comment by jessebrennan** [comment]
@joelarmstrong I think this is ready now.

**Comment by joelarmstrong** [comment]
Testing this now. It may take a few days.

File: src/toil/provisioners/clusterScaler.py [hunk 5]
```diff
@@ -126,87 +131,91 @@ def addToReservation():
...
+                            # If the job would fit, but is longer than the total node allocation
+                            # extend the node allocation
+                            elif y.nReservation == None and x == nodeReservation:
+                                # Extend the node reservation to accommodate jS
+                                y.nReservation = NodeReservation(nodeShape)
+                            reservedSuccessfully = True
+
+                        else: # Does not fit, reset
+                            x = y.nReservation
+                            t = 0
```
Comment (ejacox) [inline src/toil/provisioners/clusterScaler.py]: This could be broken up into sub functions to make it more readable.

**Comment by joelarmstrong** [comment]
Really sorry for taking so long with this. I tracked down one small bug -- I think it only triggers in very special cases -- where reservedSuccessfully could get set even though the job never fits entirely in a reservation. (It occurs when the job fits the reservation so far (from `x` to `y`), but when there are still more reservations left to go before the reservations cover the job's full runtime.)

I've pushed a fix for that (now it seems to be scaling up well). But in the process I actually found a few bugs in the original binpacking algorithm, I think--not new to this PR, though. Basically, the "wallTime" for reservations doesn't always seem to be set correctly. But since the existing binpacking algorithm has been proven to work well before (even if accidentally!), I think this PR is good to go. Any bugs in the binpacking algorithm already existed before.

I'm refactoring the binpacking a bit, to try to track down these bugs and so it's relatively testable--but that doesn't need to go in this PR, I think this is OK to merge as is.

File: src/toil/test/provisioners/clusterScalerTest.py [hunk 6]
```diff
@@ -153,47 +159,113 @@ def testClusterScaling(self):
...
+        
+        mock = MockBatchSystemAndProvisioner(config, secondsPerJob=2.0)
+        clusterScaler = ClusterScaler(mock, mock, config)
+        clusterScaler.start()
+        mock.start()
+
+        try:
+            #Add small jobs
+            map(lambda x: mock.addJob(jobShape=smallNode), range(numJobs))
+            map(lambda x: mock.addJob(jobShape=mediumNode), range(numJobs))
```
Comment (adderan) [inline src/toil/test/provisioners/clusterScalerTest.py]: This is what broke the last build, should be list(map(

Comment (ejacox) [reply src/toil/test/provisioners/clusterScalerTest.py]: OK. Thanks @adderan

**Pull request merged**